### PR TITLE
[Issue-2492] Support in-app staking for KREST

### PR DESCRIPTION
**Pull request by bluezdot**
### Related issue(s)
#2492 

### Is your feature request related to a problem(s)? Please describe.

- Support in-app staking for KREST
- Add new RPC: wss://krest.unitedbloc.com/

### Describe the solution you'd like

- Support in-app staking for KREST
- Add new RPC: wss://krest.unitedbloc.com/

### Describe alternatives you've considered

- No

### Additional context

- No

### CODE REVIEW

**Comment by saltict** [comment]
🚀 Deployed on https://pr-2503--sw-web-runner.netlify.app

File: packages/extension-base/src/services/chain-service/constants.ts [hunk 1]
```diff
@@ -60,7 +60,8 @@ export const _STAKING_CHAIN_GROUP = {
   relay: ['polkadot', 'kusama', 'aleph', 'polkadex', 'ternoa', 'ternoa_alphanet', 'alephTest', 'polkadexTest', 'westend', 'kate', 'edgeware', 'creditcoin', 'vara_network', 'goldberg_testnet'],
   para: ['moonbeam', 'moonriver', 'moonbase', 'turing', 'turingStaging', 'bifrost', 'bifrost_testnet', 'calamari_test', 'calamari', 'manta_network'],
   astar: ['astar', 'shiden', 'shibuya'],
-  amplitude: ['amplitude', 'amplitude_test', 'kilt', 'kilt_peregrine', 'pendulum'], // amplitude and kilt only share some common logic
+  amplitude: ['amplitude', 'amplitude_test', 'kilt', 'kilt_peregrine', 'pendulum', 'krest_network'], // amplitude and kilt only share some common logic
+  krest_network: ['krest_network'],
```
Comment (nampc1) [inline packages/extension-base/src/services/chain-service/constants.ts:64]: why is krest in 2 groups?

File: packages/extension-base/src/services/chain-service/constants.ts [hunk 2]
```diff
@@ -99,7 +100,8 @@ export const _STAKING_ERA_LENGTH_MAP: Record<string, number> = { // in hours
   creditcoin: 24,
   vara_network: 12,
   goldberg_testnet: 24,
-  manta_network: 6
+  manta_network: 6,
+  krest_network: 4
```
Comment (nampc1) [inline packages/extension-base/src/services/chain-service/constants.ts:104]: where's the source for this?

File: packages/extension-base/src/koni/api/staking/paraChain.ts [hunk 3]
```diff
@@ -77,6 +86,7 @@ function getSingleStakingAmplitude (substrateApi: _SubstrateApi, address: string
       } as StakingItem;
 
       stakingCallback(chain, stakingItem);
+      console.log(123123);
```
Comment (nampc1) [inline packages/extension-base/src/koni/api/staking/paraChain.ts:89]: Remove log

File: packages/extension-base/src/koni/api/staking/paraChain.ts [hunk 4]
```diff
@@ -21,8 +21,17 @@ function getSingleStakingAmplitude (substrateApi: _SubstrateApi, address: string
...
     [substrateApi.api.query.parachainStaking.unstaking, address]
   ], async ([_delegatorState, _unstaking]) => {
-    const delegatorState = _delegatorState.toPrimitive() as unknown as ParachainStakingStakeOption;
-    const unstakingInfo = _unstaking.toPrimitive() as unknown as Record<string, number>;
+    let delegatorState: ParachainStakingStakeOption;
+
+    if (_STAKING_CHAIN_GROUP.krest_network.includes(chain)) {
+      const krestDelegatorState = _delegatorState.toPrimitive() as unknown as KrestDelegateState;
+
+      delegatorState = krestDelegatorState?.delegations[0] as unknown as ParachainStakingStakeOption;
```
Comment (nampc1) [inline packages/extension-base/src/koni/api/staking/paraChain.ts:29]: It's better to iterate through all delegation record than assuming there's only 1 record

File: packages/extension-base/src/koni/api/staking/paraChain.ts [hunk 5]
```diff
@@ -50,7 +59,7 @@ function getSingleStakingAmplitude (substrateApi: _SubstrateApi, address: string
         unstakings: []
       } as NominatorMetadata);
     } else {
-      const activeBalance = delegatorState ? new BN(delegatorState.amount.toString()) : BN_ZERO;
+      const activeBalance = delegatorState ? new BN(delegatorState.amount?.toString()) : BN_ZERO;
```
Comment (nampc1) [inline packages/extension-base/src/koni/api/staking/paraChain.ts:62]: Can be done better by counting active stake and minus it from the total stake

**Review by nampc1 — CHANGES_REQUESTED** [review]

File: packages/extension-base/src/services/chain-service/constants.ts (hunk 1 above)
Comment (bluezdot) [reply packages/extension-base/src/services/chain-service/constants.ts:64]: This case is the same as amplitude-kilt relationships; amplitude and krest_network also only share some common logic. The krest_network group is to deal with specific feature on krest staking.

File: packages/extension-base/src/services/chain-service/constants.ts (hunk 2 above)
Comment (bluezdot) [reply packages/extension-base/src/services/chain-service/constants.ts:104]: the block time is: 12s
block per round is: 1200 blocks (parachainStaking.defaultBlocksPerRound)
=> round_time = 12 * 1200 / 60 / 60 = 4 hours

More Infor:
number of unlocking blocks: 50400 blocks (parachainStaking.stakeDuration)
=> unlocking time = 50400*12/60/60/24 = 7 days

File: packages/extension-base/src/koni/api/staking/paraChain.ts (hunk 4 above)
Comment (bluezdot) [reply packages/extension-base/src/koni/api/staking/paraChain.ts]: amplitude and krest are currently using the same function to subscribe nominator metadata; this handling code also forces the amplitude nominator, which is always 1 record, to be an array of one record

File: packages/extension-base/src/koni/api/staking/bonding/amplitude.ts [hunk 6]
```diff
@@ -93,18 +93,23 @@ export async function subscribeAmplitudeNominatorMetadata (chainInfo: _ChainInfo
...
+      const delegate = delegatorState[i];
+      const [identity] = identities[i];
 
       activeStake = delegate.amount.toString();
       const bnActiveStake = new BN(activeStake);
-      let delegationStatus: StakingStatus = StakingStatus.NOT_EARNING;
+      let delegationStatus = StakingStatus.NOT_EARNING;
 
-      if (bnActiveStake.gt(BN_ZERO) && bnActiveStake.gte(new BN(minDelegatorStake))) {
+      if (bnActiveStake.gte(new BN(minDelegatorStake))) {
```
Comment (nampc1) [inline packages/extension-base/src/koni/api/staking/bonding/amplitude.ts:112]: should also check active stake against zero

**Review by saltict — APPROVED** [review]

**Pull request merged**